Sources/Foundation/Decimal.swift

```diff
@@ -1687,64 +1687,74 @@ public func NSDecimalString(_ dcm: UnsafePointer<Decimal>, _ locale: Any?) -> St
         return "0"
     }
 
+    var decimalSeparator: String!
+    if let locale = locale as? Locale {
+        decimalSeparator = locale.decimalSeparator
+    } else if let dictionary = locale as? [NSLocale.Key: String] {
+        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator]
+    } else if let dictionary = locale as? [String: String] {
+        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator.rawValue]
+    } else {
+        decimalSeparator = "."
+    }
+    let decimalSeparatorCChar = decimalSeparator.utf8CString // should be accessed with withUnsafeBufferPointer
+    let decimalSeparatorCCharCount = decimalSeparatorCChar.withUnsafeBufferPointer { $0.count } - 1 // without null terminator
+
     let ZERO: CChar = 0x30 // ASCII '0' == 0x30
-    let DECIMALPOINT: CChar = 0x2e // ASCII '.' == 0x2e
     let MINUS: CChar = 0x2d // ASCII '-' == 0x2d
-    let bufferSize = 200 // Max value: 39 + 128 + sign + decimal point
-    var buffer = Array<CChar>(repeating: 0, count: bufferSize)
+    // FIXME: where magic constants 39 and 128 come from?
+    let bufferSize = 39 + 128 + 1 /* sign */ + decimalSeparatorCCharCount + 1 /* null terminator */
```

Comment:
The 39 is from the mantissa being an 128bit, so log10(2^128) is approx 38.5 giving 39 digits max for the mantissa.
128 is the exponent being an `Int8`